src/main/java/org/mal_lang/langspec/LangReader.java

```diff
@@ -0,0 +1,115 @@
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.zip.ZipInputStream;
+
+/** Reads {@link Lang} objects from {@code .mar} files. */
+public final class LangReader {
+  private final InputStream in;
+  private JsonObject langSpec = null;
+  private final Map<String, byte[]> icons = new LinkedHashMap<>();
+  private String license = null;
+  private String notice = null;
+
+  /**
+   * Constructs a new {@code LangReader} object.
+   *
+   * @param in an input stream of a {@code .mar} file
+   * @throws NullPointerException if {@code in} is {@code null}
+   */
+  public LangReader(InputStream in) {
+    checkNotNull(in);
+    this.in = in;
+  }
+
+  private void readLangSpec(ZipInputStream zipIn) throws IOException {
+    var jsonReaderFactory = Json.createReaderFactory(null);
+    var jsonReader = jsonReaderFactory.createReader(zipIn, StandardCharsets.UTF_8);
+    try {
+      this.langSpec = jsonReader.readObject();
+    } catch (JsonParsingException e) {
+      throw new IllegalArgumentException("Failed to parse \"langspec.json\" as a JSON object", e);
+    } catch (JsonException e) {
+      var cause = e.getCause();
+      if (cause == null || !(cause instanceof IOException)) {
+        throw new RuntimeException("JsonException did not contain an IOException cause");
+      }
+      throw (IOException) cause;
+    }
+  }
+
+  private void readIcon(ZipInputStream zipIn, String name) throws IOException {
+    // Remove "lang/" prefix
+    var fileName = name.substring(6);
+    if (!fileName.endsWith(".png") && !fileName.endsWith(".svg")) {
+      return;
+    }
+    // Remove ".svg" or ".png" suffix
+    var assetName = fileName.substring(0, fileName.length() - 4);
+    if (!Utils.isIdentifier(assetName)) {
+      return;
+    }
+    this.icons.put(fileName, zipIn.readAllBytes());
+  }
+
+  private String readString(ZipInputStream zipIn) throws IOException {
+    return new String(zipIn.readAllBytes(), StandardCharsets.UTF_8);
+  }
+
+  /**
+   * Reads a {@link Lang} object from the input stream.
+   *
+   * @return a {@link Lang} object from the input stream
+   * @throws IOException if an I/O error has occurred
+   * @throws IllegalArgumentException if the input stream contains invalid data
+   */
+  public Lang read() throws IOException {
+    try (var zipIn = new ZipInputStream(in)) {
+      for (var zipEntry = zipIn.getNextEntry(); zipEntry != null; zipEntry = zipIn.getNextEntry()) {
+        if (!zipEntry.isDirectory()) {
+          var name = zipEntry.getName();
+          if (name.equals("langspec.json")) {
+            readLangSpec(zipIn);
+          } else if (name.startsWith("icons/")) {
+            readIcon(zipIn, name);
+          } else if (name.equals("LICENSE")) {
+            this.license = readString(zipIn);
+          } else if (name.equals("NOTICE")) {
+            this.notice = readString(zipIn);
+          }
+        }
+        zipIn.closeEntry();
```

Comment:
Hm, I think java now has some form of open with that gives auto close

Reply:
zip files are special. `zipIn` is a `ZipInputStream` that is closed at the end of scope on line 109. `closeEntry()` `Closes the current ZIP entry and positions the stream for reading the  next entry`. It seems like `getNextEntry()` that `Reads the next ZIP file entry and positions the stream at the beginning of the entry data` on line 94 calls `closeEntry()`, so I'll remove this.